libretro/libretro.cpp

```diff
@@ -379,9 +386,112 @@ static RetroOption<bool> ppsspp_retain_changed_textures("ppsspp_retain_changed_t
+    {"socom.cc", "socom.cc"},
+    {"myneighborsushicat.com", "myneighborsushicat.com"},
+    {"localhost", "localhost"},
+    {"IP address", "IP address"}
+});
+static RetroOption<int> ppsspp_pro_ad_hoc_ipv4[] = {
+   {"ppsspp_pro_ad_hoc_server_address01", "PRO ad hoc server IP address Pt  1: x--.---.---.--- ", 0, 10, 1},
+   {"ppsspp_pro_ad_hoc_server_address02", "PRO ad hoc server IP address Pt  2: -x-.---.---.--- ", 0, 10, 1},
+   {"ppsspp_pro_ad_hoc_server_address03", "PRO ad hoc server IP address Pt  3: --x.---.---.--- ", 0, 10, 1},
+   {"ppsspp_pro_ad_hoc_server_address04", "PRO ad hoc server IP address Pt  4: ---.x--.---.--- ", 0, 10, 1},
+   {"ppsspp_pro_ad_hoc_server_address05", "PRO ad hoc server IP address Pt  5: ---.-x-.---.--- ", 0, 10, 1},
+   {"ppsspp_pro_ad_hoc_server_address06", "PRO ad hoc server IP address Pt  6: ---.--x.---.--- ", 0, 10, 1},
+   {"ppsspp_pro_ad_hoc_server_address07", "PRO ad hoc server IP address Pt  7: ---.---.x--.--- ", 0, 10, 1},
+   {"ppsspp_pro_ad_hoc_server_address08", "PRO ad hoc server IP address Pt  8: ---.---.-x-.--- ", 0, 10, 1},
+   {"ppsspp_pro_ad_hoc_server_address09", "PRO ad hoc server IP address Pt  9: ---.---.--x.--- ", 0, 10, 1},
+   {"ppsspp_pro_ad_hoc_server_address10", "PRO ad hoc server IP address Pt 10: ---.---.---.x-- ", 0, 10, 1},
+   {"ppsspp_pro_ad_hoc_server_address11", "PRO ad hoc server IP address Pt 11: ---.---.---.-x- ", 0, 10, 1},
+   {"ppsspp_pro_ad_hoc_server_address12", "PRO ad hoc server IP address Pt 12: ---.---.---.--x ", 0, 10, 1}
+};
+static RetroOption<bool> ppsspp_enable_upnp("ppsspp_enable_upnp", "Enable UPnP (need a few seconds to detect)", true);
+static RetroOption<bool> ppsspp_upnp_use_original_port("ppsspp_upnp_use_original_port", "UPnP use original port (enabled = PSP compatibility)", true);
+static RetroOption<bool> ppsspp_enable_network_chat("ppsspp_enable_network_chat", "Enable network chat", true);
+static RetroOption<int> ppsspp_chat_button_position("ppsspp_chat_button_position", "Chat button position", {
+    {"Bottom left", BOTTOM_LEFT},
+    {"Bottom center", BOTTOM_CENTER},
+    {"Bottom right", BOTOM_RIGHT},
+    {"Top left", TOP_LEFT},
+    {"Top center", TOP_CENTER},
+    {"Top right", TOP_RIGHT},
+    {"Center left", CENTER_LEFT},
+    {"Center right", CENTER_RIGHT},
+    {"None", 8}
+});
+static RetroOption<int> ppsspp_chat_screen_position("ppsspp_chat_screen_position", "Chat screen position", {
+    {"Bottom left", BOTTOM_LEFT},
+    {"Bottom center", BOTTOM_CENTER},
+    {"Bottom right", BOTOM_RIGHT},
+    {"Top left", TOP_LEFT},
+    {"Top center", TOP_CENTER},
+    {"Top right", TOP_RIGHT},
+    {"Center left", CENTER_LEFT},
+    {"Center right", CENTER_RIGHT}
+});
+static RetroOption<int> ppsspp_port_offset("ppsspp_port_offset", "Port offset (0 = PSP compatibility)", {{"0", 0}, {"5000", 5000}, {"10000", 10000}, {"15000", 15000}});
```

Comment:
Is this limiting port offset to 0, 5000, 10000, 15000 ? 
If so, this will make libretro unable to play with standalone version that use free value (just like Minimum timeout setting)

Reply:
Yes indeed.

My first version had a drop down with 65535 possible choices. That took quite long to open, even on a fast desktop. And it was quite unpractical to scroll all the way down to 10000 (which is the value suggested in the AdHoc wiki). I ended up opening the options file with a text editor after ssh-ing into my Pi.

The Libretro core targets devices without a keyboard. As far as I can see, either we supply a discrete list of sane choices, or we use a single drop down control for each digit of the port offset  (similar to the IP address).

I was hoping we could get away with the discrete list of sane choices in this case.

Reply:
Would it be acceptable if I extend the list to 0, 5000-5020, 10000-10020, 15000-15020?

Reply:
Why not making it like the Minimum Timeout? it seems to use Min, Max, and Step, right?

If libretro doesn't support using Min,Max,Step, may be @MojoJojoDojo can tell us which port offset are commonly used by players? i heard some ppl also use 65000?

Reply:
Well the default is 10000 so most people use that and it works fine.

Reply:
did you said that some games worked better with 65k long ago?

Reply:
No other people said that 60k is a good choice for Gran Turismo but besides that I dont think i managed to get it to work well...
Previously I mostly used offset 1500 for playing though.

Reply:
> Why not making it like the Minimum Timeout? which seems to use Min, Max, and Step, right?
> 
> If libretro doesn't support using Min,Max,Step, may be @MojoJojoDojo can tell us which port offset are commonly used by players? i heard some ppl also use 65000?

Using Min, Max, and Step to generate a drop down with entries between 0 and 65535 is definitely possible, technically. But it does not yield a usable input widget. If in doubt, just change the constructor of the ppsspp_port_offset RetroOption to
`
static RetroOption<int> ppsspp_port_offset("ppsspp_port_offset", "Port offset (0 = PSP compatibility)", 0, 65535, 1);
`
and see for yourself. It would be much better if we can settle for a sane and limited set of choices. 

As long as people who want to play a network game together can agree on a common port to use, we should be good?

Reply:
the step doesn't need to be 1, it can be 500 or 1k or even larger steps

Reply:
Thank you for the explanation, 1k steps would certainly yield a sane and discrete set of options.